Convert to Elm 0.17 (WIP)

stkb opened the pull request:
This is still a WIP and I haven't touched the documentation yet.

Html in Elm 0.17 now deals with `Html msg` and `Attribute msg` in place of just `Html` and `Attribute`, so the conversion had consisted mainly of adding these extra msg parameters everywhere, and converting all instances of Signal.Message to msg instead.

An issue I had is with the events in Html.Shorthand.Event: onInput, onChange, onEnter, onKeyboardLost, onMouseLost. These all did (and still do) take a decoder and a message function: `Json.Decoder a -> (a -> msg) -> Attribute msg`, similar to the old Html.Events.on function: `String -> Decoder a -> (a -> Message) -> Attribute`.

However in 0.17 the Html.Events.on function now has a simplified signature: `String -> Decoder msg -> Attribute msg`, and I'm not sure whether these events should follow suit or not, and what the consequences would be.

Additionally, the new Html.Events now has its own onInput event, so I've renamed the one here to `onInput'` for now, but I'm not sure what the best name is.

stkb committed:
Convert to Elm 0.17 (WIP)

rehno-lindeque commented:
Fantastic, thank you for this.

> However in 0.17 the Html.Events.on function now has a simplified signature: String -> Decoder msg -> Attribute msg, and I'm not sure whether these events should follow suit or not, and what the consequences would be.

Regarding these, it would be nice to try and mimic what elm-html does here since this package is intended to be "just" a shorthand for elm-html. That said, I'm not exactly sure about the consequences either (it's possible that this allows for some extra filtering for events). I'd say let's update these to the new style, but perhaps just as a separate commit in order to have a bit of a history when upgrading dependent code.

> Additionally, the new Html.Events now has its own onInput event, so I've renamed the one here to onInput' for now, but I'm not sure what the best name is.

I'll quickly take a look at this tomorrow - I'm not sure about this one right now either. `onInput'` sounds like a good compromise in the mean time

stkb committed:
Update documentation

messageDecoder: Removed note about sending to different channels since that's
not applicable any more.

stkb commented:
I've added another commit that updates the documentation.

@rehno-lindeque Thanks for looking at this.

About the onInput event, what it effectively now is, is the same as the onKeyboardLost and onMouseLost events, i.e: a copy of an existing event in the core libraries (onBlur & onMouseLeave respectively), but with access to the full event information through a decoder.

Therefore the name `onInput'` doesn't really fit with the other two, but I can't think of a better name right now either.

Maybe it's not even needed at all, if no one ever used anything other than the targetValue decoder with it.

I don't have an existing codebase to be able to answer these questions, so I'll leave this here.

rehno-lindeque commented:
Thanks @stkb, I've merged and given you commit rights to elm html packages in case there's still anything lacking. It's published as http://package.elm-lang.org/packages/circuithub/elm-html-shorthand/11.0.0/

stkb commented:
@rehno-lindeque Thanks!
